File: tests/test_cluster_centroid.cpp

#include "CLUEstering/CLUEstering.hpp"

#include <cmath>
#include <ranges>
#include <span>
#include <vector>

#include <fmt/core.h>
#include <fmt/format.h>

#define DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
#include "doctest.h"

TEST_CASE("Test computation of cluster centroid") {
  const auto device = clue::get_device(0u);
  clue::Queue queue(device);

  const auto test_file_path = std::string(TEST_DATA_DIR) + "/data_32768.csv";
  clue::PointsHost<2> h_points = clue::read_csv<2, float>(queue, test_file_path);
  const auto n_points = h_points.size();
  clue::PointsDevice<2> d_points(queue, n_points);

  const float dc{21.f}, rhoc{10.f}, outlier{21.f};
  clue::Clusterer<2> algo(queue, dc, rhoc, outlier);

  algo.make_clusters(queue, h_points, d_points);

  SUBCASE("Check centroid of cluster 0") {
    auto centroid = clue::cluster_centroid(h_points, 0);
    std::ranges::for_each(centroid, [](auto coord) -> void {
      CHECK(std::isfinite(coord));
      CHECK(!std::isnan(coord));
    });
  }
}

TEST_CASE("Test computation of all cluster centroids") {
  const auto device = clue::get_device(0u);
  clue::Queue queue(device);

  const auto test_file_path = std::string(TEST_DATA_DIR) + "/data_32768.csv";
  clue::PointsHost<2> h_points = clue::read_csv<2, float>(queue, test_file_path);
  const auto n_points = h_points.size();
  clue::PointsDevice<2> d_points(queue, n_points);

  const float dc{21.f}, rhoc{10.f}, outlier{21.f};
  clue::Clusterer<2> algo(queue, dc, rhoc, outlier);

  algo.make_clusters(queue, h_points, d_points);
  auto centroids = clue::cluster_centroids(h_points);
  auto clusters = clue::get_clusters(h_points);
  CHECK(centroids.size() == clusters.size());
}

File: tests/test_cluster_centroid.py
'''
Test clusterer methods for computing cluster centroids
'''

import os
import sys
import pandas as pd
import pytest
from check_result import check_result
sys.path.insert(1, '../CLUEstering/')
import CLUEstering as clue

@pytest.fixture
def dataset():
    '''
    Returns the dataframe containing the sissa dataset 1000 points
    '''
    return pd.read_csv("../data/sissa_1000.csv")

def test_cluster_centroids(dataset):
    '''
    Test the methods for computing cluster centroids
    '''
    c = clue.clusterer(21., 10., 21.)
    c.read_data(dataset)
    c.run_clue()

    centroids = c.cluster_centroids()
    assert centroids.shape == (c.n_clusters, c.n_dim)
    single_centroids = []
    for cluster in range(c.n_clusters):
        print(cluster)
        centroid = c.cluster_centroid(cluster)
        single_centroids.append(centroid)

    assert (single_centroids == centroids).all()
